bigquery/src/main/java/org/apache/iceberg/gcp/bigquery/BigQueryMetastoreCatalog.java

```diff
@@ -0,0 +1,370 @@
+import org.apache.iceberg.catalog.TableIdentifier;
+import org.apache.iceberg.exceptions.NoSuchNamespaceException;
+import org.apache.iceberg.exceptions.NoSuchTableException;
+import org.apache.iceberg.exceptions.ServiceFailureException;
+import org.apache.iceberg.exceptions.ValidationException;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.relocated.com.google.common.annotations.VisibleForTesting;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.base.Strings;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.util.LocationUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Iceberg Bigquery Metastore Catalog implementation. */
+public class BigQueryMetastoreCatalog extends BaseMetastoreCatalog
+    implements SupportsNamespaces, Configurable {
+
+  // User provided properties.
+  public static final String PROJECT_ID = "gcp.bigquery.project-id";
+  public static final String GCP_LOCATION = "gcp.bigquery.location";
+  public static final String LIST_ALL_TABLES = "gcp.bigquery.list-all-tables";
+
+  private static final Logger LOG = LoggerFactory.getLogger(BigQueryMetastoreCatalog.class);
+
+  private static final String DEFAULT_GCP_LOCATION = "us";
+
+  private String catalogName;
+  private Map<String, String> catalogProperties;
+  private FileIO fileIO;
+  private Configuration conf;
+  private String projectId;
+  private String projectLocation;
+  private BigQueryMetastoreClient client;
+  private boolean listAllTables;
+  private String warehouseLocation;
+
+  // Must have a no-arg constructor to be dynamically loaded
+  // initialize(String name, Map<String, String> properties) will be called to complete
+  // initialization
+  public BigQueryMetastoreCatalog() {}
+
+  @Override
+  public void initialize(String name, Map<String, String> properties) {
+    Preconditions.checkArgument(
+        properties.containsKey(PROJECT_ID),
+        "Invalid GCP project: %s must be specified",
+        PROJECT_ID);
+
+    this.projectId = properties.get(PROJECT_ID);
+    this.projectLocation = properties.getOrDefault(GCP_LOCATION, DEFAULT_GCP_LOCATION);
+
+    BigQueryOptions options =
+        BigQueryOptions.newBuilder()
+            .setProjectId(projectId)
+            .setLocation(projectLocation)
+            .setRetrySettings(ServiceOptions.getDefaultRetrySettings())
+            .build();
+
+    try {
+      client = new BigQueryMetastoreClientImpl(options);
+    } catch (IOException e) {
+      throw new ServiceFailureException(e, "Creating BigQuery client failed");
+    } catch (GeneralSecurityException e) {
+      throw new ValidationException(e, "Creating BigQuery client failed due to a security issue");
```

Comment:
This is also not the right exception to throw, if this a required exception we probably need to find a better fit (or just wrap in RuntimeException).

Reply:
I wrapped it with RTE